# Frameworks


In [63]:
import pandas as pd
import plotly.express as px

# Load Dataset

In [64]:
##Load Data
df=pd.read_csv("amazon.csv")
df.head()

,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count,about_product,user_id,user_name,review_id,review_title,review_content,img_link,product_link
0,B07JW9H4J1,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories|Accessories&Peripherals|...,₹399,"₹1,099",64%,4.2,"24,269",High Compatibility : Compatible With iPhone 12...,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...","R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...","Satisfied,Charging is really fast,Value for mo...",Looks durable Charging is fine tooNo complains...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...
1,B098NS6PVG,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories|Accessories&Peripherals|...,₹199,₹349,43%,4.0,"43,994","Compatible with all Type C enabled devices, be...","AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...","ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...","RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...","A Good Braided Cable for Your Type C Device,Go...",I ordered this cable to connect my phone to An...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Ambrane-Unbreakable-Char...
2,B096MSW6CT,Sounce Fast Phone Charging Cable & Data Sync U...,Computers&Accessories|Accessories&Peripherals|...,₹199,"₹1,899",90%,3.9,"7,928",【 Fast Charger& Data Sync】-With built-in safet...,"AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...","Kunal,Himanshu,viswanath,sai niharka,saqib mal...","R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...","Good speed for earlier versions,Good Product,W...","Not quite durable and sturdy,https://m.media-a...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Sounce-iPhone-Charging-C...
3,B08HDJ86NZ,boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...,Computers&Accessories|Accessories&Peripherals|...,₹329,₹699,53%,4.2,"94,363",The boAt Deuce USB 300 2 in 1 cable is compati...,"AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...","Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...","R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...","Good product,Good one,Nice,Really nice product...","Good product,long wire,Charges good,Nice,I bou...",https://m.media-amazon.com/images/I/41V5FtEWPk...,https://www.amazon.in/Deuce-300-Resistant-Tang...
4,B08CF3B7N1,Portronics Konnect L 1.2M Fast Charging 3A 8 P...,Computers&Accessories|Accessories&Peripherals|...,₹154,₹399,61%,4.2,"16,905",[CHARGE & SYNC FUNCTION]- This cable comes wit...,"AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...","rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...","R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...","As good as original,Decent,Good one for second...","Bought this instead of original apple, does th...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Portronics-Konnect-POR-1...


# Dataset Overview


In [65]:
print("Data Shape",df.shape)
print("Size of data",df.size)

Data Shape (1465, 16)
Size of data 23440


In [66]:
print(df.columns)

Index(['product_id', 'product_name', 'category', 'discounted_price',
       'actual_price', 'discount_percentage', 'rating', 'rating_count',
       'about_product', 'user_id', 'user_name', 'review_id', 'review_title',
       'review_content', 'img_link', 'product_link'],
      dtype='object')


In [67]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   product_id           1465 non-null   object
 1   product_name         1465 non-null   object
 2   category             1465 non-null   object
 3   discounted_price     1465 non-null   object
 4   actual_price         1465 non-null   object
 5   discount_percentage  1465 non-null   object
 6   rating               1465 non-null   object
 7   rating_count         1463 non-null   object
 8   about_product        1465 non-null   object
 9   user_id              1465 non-null   object
 10  user_name            1465 non-null   object
 11  review_id            1465 non-null   object
 12  review_title         1465 non-null   object
 13  review_content       1465 non-null   object
 14  img_link             1465 non-null   object
 15  product_link         1465 non-null   object
dtypes: obj

# Data Cleanning

## Drop Un-Wanted Columns

In [68]:
df.drop(
    columns=[
        "img_link",
        "product_link",
        "review_content",
        "review_title",
        "review_id",
        "user_name",
        "user_id",
        "about_product"
    ],inplace=True
)

In [69]:
df.isnull().sum()

product_id             0
product_name           0
category               0
discounted_price       0
actual_price           0
discount_percentage    0
rating                 0
rating_count           2
dtype: int64

In [70]:
df[df["rating_count"].isnull()]

,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count
282,B0B94JPY2N,Amazon Brand - Solimo 65W Fast Charging Braide...,Computers&Accessories|Accessories&Peripherals|...,₹199,₹999,80%,3.0,NaN
324,B0BQRJ3C47,"REDTECH USB-C to Lightning Cable 3.3FT, [Apple...",Computers&Accessories|Accessories&Peripherals|...,₹249,₹999,75%,5.0,NaN


In [71]:
df = df.dropna()

In [72]:
df.shape

(1463, 8)

In [73]:
df.isnull().sum()

product_id             0
product_name           0
category               0
discounted_price       0
actual_price           0
discount_percentage    0
rating                 0
rating_count           0
dtype: int64

## Fix data types

##### Datatype change Required

1.discounted_price->float

2.actual_price->float

3.discount_percentage->int

4.rating->float
5.rating_count->int

In [74]:
# Price
price_columns = ["discounted_price", "actual_price"]

df[price_columns] = (
    df[price_columns]
    .replace({"₹": "", ",": ""}, regex=True)
    .astype(float)
)

print(df)

      product_id                                       product_name  \
0     B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...   
1     B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2     B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3     B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4     B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   
...          ...                                                ...   
1460  B08L7J3T31  Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...   
1461  B01M6453MB  Prestige Delight PRWO Electric Rice Cooker (1 ...   
1462  B009P2LIL4  Bajaj Majesty RX10 2000 Watts Heat Convector R...   
1463  B00J5DYCCA  Havells Ventil Air DSP 230mm Exhaust Fan (Pist...   
1464  B01486F4G6  Borosil Jumbo 1000-Watt Grill Sandwich Maker (...   

                                               category  discounted_price  \
0     Computers&Accessories|Accessories&Peripherals|...             39

In [75]:
df

,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count
0,B07JW9H4J1,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories|Accessories&Peripherals|...,399.0,1099.0,64%,4.2,"24,269"
1,B098NS6PVG,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories|Accessories&Peripherals|...,199.0,349.0,43%,4.0,"43,994"
2,B096MSW6CT,Sounce Fast Phone Charging Cable & Data Sync U...,Computers&Accessories|Accessories&Peripherals|...,199.0,1899.0,90%,3.9,"7,928"
3,B08HDJ86NZ,boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...,Computers&Accessories|Accessories&Peripherals|...,329.0,699.0,53%,4.2,"94,363"
4,B08CF3B7N1,Portronics Konnect L 1.2M Fast Charging 3A 8 P...,Computers&Accessories|Accessories&Peripherals|...,154.0,399.0,61%,4.2,"16,905"
...,...,...,...,...,...,...,...,...
1460,B08L7J3T31,Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...,Home&Kitchen|Kitchen&HomeAppliances|WaterPurif...,379.0,919.0,59%,4,"1,090"
1461,B01M6453MB,Prestige Delight PRWO Electric Rice Cooker (1 ...,Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...,2280.0,3045.0,25%,4.1,"4,118"
1462,B009P2LIL4,Bajaj Majesty RX10 2000 Watts Heat Convector R...,"Home&Kitchen|Heating,Cooling&AirQuality|RoomHe...",2219.0,3080.0,28%,3.6,468
1463,B00J5DYCCA,Havells Ventil Air DSP 230mm Exhaust Fan (Pist...,"Home&Kitchen|Heating,Cooling&AirQuality|Fans|E...",1399.0,1890.0,26%,4,"8,031"


In [76]:
df.discount_percentage = (
    df.discount_percentage
    .replace({"%": "", ",": ""}, regex=True)
    .astype(float)
)

In [77]:
df

,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count
0,B07JW9H4J1,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories|Accessories&Peripherals|...,399.0,1099.0,64.0,4.2,"24,269"
1,B098NS6PVG,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories|Accessories&Peripherals|...,199.0,349.0,43.0,4.0,"43,994"
2,B096MSW6CT,Sounce Fast Phone Charging Cable & Data Sync U...,Computers&Accessories|Accessories&Peripherals|...,199.0,1899.0,90.0,3.9,"7,928"
3,B08HDJ86NZ,boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...,Computers&Accessories|Accessories&Peripherals|...,329.0,699.0,53.0,4.2,"94,363"
4,B08CF3B7N1,Portronics Konnect L 1.2M Fast Charging 3A 8 P...,Computers&Accessories|Accessories&Peripherals|...,154.0,399.0,61.0,4.2,"16,905"
...,...,...,...,...,...,...,...,...
1460,B08L7J3T31,Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...,Home&Kitchen|Kitchen&HomeAppliances|WaterPurif...,379.0,919.0,59.0,4,"1,090"
1461,B01M6453MB,Prestige Delight PRWO Electric Rice Cooker (1 ...,Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...,2280.0,3045.0,25.0,4.1,"4,118"
1462,B009P2LIL4,Bajaj Majesty RX10 2000 Watts Heat Convector R...,"Home&Kitchen|Heating,Cooling&AirQuality|RoomHe...",2219.0,3080.0,28.0,3.6,468
1463,B00J5DYCCA,Havells Ventil Air DSP 230mm Exhaust Fan (Pist...,"Home&Kitchen|Heating,Cooling&AirQuality|Fans|E...",1399.0,1890.0,26.0,4,"8,031"


In [78]:
df["rating"] = pd.to_numeric(df["rating"], errors="coerce")

In [79]:
df.rating_count = (
    df.rating_count
    .replace({"%": "", ",": ""}, regex=True)
    .astype(float)
)

In [80]:
df

,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count
0,B07JW9H4J1,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories|Accessories&Peripherals|...,399.0,1099.0,64.0,4.2,24269.0
1,B098NS6PVG,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories|Accessories&Peripherals|...,199.0,349.0,43.0,4.0,43994.0
2,B096MSW6CT,Sounce Fast Phone Charging Cable & Data Sync U...,Computers&Accessories|Accessories&Peripherals|...,199.0,1899.0,90.0,3.9,7928.0
3,B08HDJ86NZ,boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...,Computers&Accessories|Accessories&Peripherals|...,329.0,699.0,53.0,4.2,94363.0
4,B08CF3B7N1,Portronics Konnect L 1.2M Fast Charging 3A 8 P...,Computers&Accessories|Accessories&Peripherals|...,154.0,399.0,61.0,4.2,16905.0
...,...,...,...,...,...,...,...,...
1460,B08L7J3T31,Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...,Home&Kitchen|Kitchen&HomeAppliances|WaterPurif...,379.0,919.0,59.0,4.0,1090.0
1461,B01M6453MB,Prestige Delight PRWO Electric Rice Cooker (1 ...,Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...,2280.0,3045.0,25.0,4.1,4118.0
1462,B009P2LIL4,Bajaj Majesty RX10 2000 Watts Heat Convector R...,"Home&Kitchen|Heating,Cooling&AirQuality|RoomHe...",2219.0,3080.0,28.0,3.6,468.0
1463,B00J5DYCCA,Havells Ventil Air DSP 230mm Exhaust Fan (Pist...,"Home&Kitchen|Heating,Cooling&AirQuality|Fans|E...",1399.0,1890.0,26.0,4.0,8031.0


In [81]:
df.isnull().sum()

product_id             0
product_name           0
category               0
discounted_price       0
actual_price           0
discount_percentage    0
rating                 1
rating_count           0
dtype: int64

In [82]:
df = df.dropna()

## string parsing/splitting

In [83]:
## Category fix
df["category"] = df["category"].str.split("|").str[0]

C:\Users\Sahil Khan\AppData\Local\Temp\ipykernel_22980\454582929.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [84]:
df

,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count
0,B07JW9H4J1,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories,399.0,1099.0,64.0,4.2,24269.0
1,B098NS6PVG,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories,199.0,349.0,43.0,4.0,43994.0
2,B096MSW6CT,Sounce Fast Phone Charging Cable & Data Sync U...,Computers&Accessories,199.0,1899.0,90.0,3.9,7928.0
3,B08HDJ86NZ,boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...,Computers&Accessories,329.0,699.0,53.0,4.2,94363.0
4,B08CF3B7N1,Portronics Konnect L 1.2M Fast Charging 3A 8 P...,Computers&Accessories,154.0,399.0,61.0,4.2,16905.0
...,...,...,...,...,...,...,...,...
1460,B08L7J3T31,Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...,Home&Kitchen,379.0,919.0,59.0,4.0,1090.0
1461,B01M6453MB,Prestige Delight PRWO Electric Rice Cooker (1 ...,Home&Kitchen,2280.0,3045.0,25.0,4.1,4118.0
1462,B009P2LIL4,Bajaj Majesty RX10 2000 Watts Heat Convector R...,Home&Kitchen,2219.0,3080.0,28.0,3.6,468.0
1463,B00J5DYCCA,Havells Ventil Air DSP 230mm Exhaust Fan (Pist...,Home&Kitchen,1399.0,1890.0,26.0,4.0,8031.0


# Data Transformation

## Feature engineering

In [85]:
df["Revenue_Potential"]=df["discounted_price"]*df["rating_count"]

C:\Users\Sahil Khan\AppData\Local\Temp\ipykernel_22980\1563695500.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [86]:
df

,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count,Revenue_Potential
0,B07JW9H4J1,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories,399.0,1099.0,64.0,4.2,24269.0,9683331.0
1,B098NS6PVG,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories,199.0,349.0,43.0,4.0,43994.0,8754806.0
2,B096MSW6CT,Sounce Fast Phone Charging Cable & Data Sync U...,Computers&Accessories,199.0,1899.0,90.0,3.9,7928.0,1577672.0
3,B08HDJ86NZ,boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...,Computers&Accessories,329.0,699.0,53.0,4.2,94363.0,31045427.0
4,B08CF3B7N1,Portronics Konnect L 1.2M Fast Charging 3A 8 P...,Computers&Accessories,154.0,399.0,61.0,4.2,16905.0,2603370.0
...,...,...,...,...,...,...,...,...,...
1460,B08L7J3T31,Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...,Home&Kitchen,379.0,919.0,59.0,4.0,1090.0,413110.0
1461,B01M6453MB,Prestige Delight PRWO Electric Rice Cooker (1 ...,Home&Kitchen,2280.0,3045.0,25.0,4.1,4118.0,9389040.0
1462,B009P2LIL4,Bajaj Majesty RX10 2000 Watts Heat Convector R...,Home&Kitchen,2219.0,3080.0,28.0,3.6,468.0,1038492.0
1463,B00J5DYCCA,Havells Ventil Air DSP 230mm Exhaust Fan (Pist...,Home&Kitchen,1399.0,1890.0,26.0,4.0,8031.0,11235369.0


In [87]:
df['absolute_discounted_amount']=df['actual_price']-df['discounted_price']

C:\Users\Sahil Khan\AppData\Local\Temp\ipykernel_22980\1819317483.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [88]:
df

,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count,Revenue_Potential,absolute_discounted_amount
0,B07JW9H4J1,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories,399.0,1099.0,64.0,4.2,24269.0,9683331.0,700.0
1,B098NS6PVG,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories,199.0,349.0,43.0,4.0,43994.0,8754806.0,150.0
2,B096MSW6CT,Sounce Fast Phone Charging Cable & Data Sync U...,Computers&Accessories,199.0,1899.0,90.0,3.9,7928.0,1577672.0,1700.0
3,B08HDJ86NZ,boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...,Computers&Accessories,329.0,699.0,53.0,4.2,94363.0,31045427.0,370.0
4,B08CF3B7N1,Portronics Konnect L 1.2M Fast Charging 3A 8 P...,Computers&Accessories,154.0,399.0,61.0,4.2,16905.0,2603370.0,245.0
...,...,...,...,...,...,...,...,...,...,...
1460,B08L7J3T31,Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...,Home&Kitchen,379.0,919.0,59.0,4.0,1090.0,413110.0,540.0
1461,B01M6453MB,Prestige Delight PRWO Electric Rice Cooker (1 ...,Home&Kitchen,2280.0,3045.0,25.0,4.1,4118.0,9389040.0,765.0
1462,B009P2LIL4,Bajaj Majesty RX10 2000 Watts Heat Convector R...,Home&Kitchen,2219.0,3080.0,28.0,3.6,468.0,1038492.0,861.0
1463,B00J5DYCCA,Havells Ventil Air DSP 230mm Exhaust Fan (Pist...,Home&Kitchen,1399.0,1890.0,26.0,4.0,8031.0,11235369.0,491.0


# Analysis

#### CO-Realtion

In [89]:
num_cols = [
    "discounted_price",
    "actual_price",
    "discount_percentage",
    "rating",
    "rating_count",
    "Revenue_Potential",
    "absolute_discounted_amount"
]

In [90]:
import plotly.express as px

fig = px.scatter_matrix(
    df,
    dimensions=num_cols,
    title="Scatter Matrix of Numerical Features"
)

fig.update_layout(
    width=1500,    # Width in pixels
    height=1500    # Height in pixels
)

fig.show()

In [91]:
corr = df[num_cols].corr()

corr

,discounted_price,actual_price,discount_percentage,rating,rating_count,Revenue_Potential,absolute_discounted_amount
discounted_price,1.000000,0.961910,-0.242298,0.121132,-0.027304,0.468167,0.763650
actual_price,0.961910,1.000000,-0.117855,0.122467,-0.036215,0.413662,0.911056
discount_percentage,-0.242298,-0.117855,1.000000,-0.155679,0.011294,-0.193979,0.087079
rating,0.121132,0.122467,-0.155679,1.000000,0.102235,0.068720,0.106554
rating_count,-0.027304,-0.036215,0.011294,0.102235,1.000000,0.416633,-0.044354
Revenue_Potential,0.468167,0.413662,-0.193979,0.068720,0.416633,1.000000,0.270902
absolute_discounted_amount,0.763650,0.911056,0.087079,0.106554,-0.044354,0.270902,1.000000


In [92]:
# Finding
#.If actual price is more then there is more discounted price
#.Higher The actual price higher the discounted amount
#.The priduct which are in higher disount people buy them more


#### 1. Pricing Efficiency & Discount Strategy

###### 1.Are high discounts actually driving meaningful volume?

In [93]:
import plotly.express as px

fig = px.scatter(
    df,
    x="discount_percentage",
    y="rating_count",
    trendline="ols",
    title="Rating Count vs Discounted Price"
)

fig.show()

#### I suggests that higher discounts are not associated with proportionally higher customer engagement (using rating count as a proxy for demand). Therefore, discounting alone does not appear to be an effective strategy for increasing demand.

In [94]:
df.head()

,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count,Revenue_Potential,absolute_discounted_amount
0,B07JW9H4J1,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories,399.0,1099.0,64.0,4.2,24269.0,9683331.0,700.0
1,B098NS6PVG,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories,199.0,349.0,43.0,4.0,43994.0,8754806.0,150.0
2,B096MSW6CT,Sounce Fast Phone Charging Cable & Data Sync U...,Computers&Accessories,199.0,1899.0,90.0,3.9,7928.0,1577672.0,1700.0
3,B08HDJ86NZ,boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...,Computers&Accessories,329.0,699.0,53.0,4.2,94363.0,31045427.0,370.0
4,B08CF3B7N1,Portronics Konnect L 1.2M Fast Charging 3A 8 P...,Computers&Accessories,154.0,399.0,61.0,4.2,16905.0,2603370.0,245.0


#### 2.Deep vs. Shallow Discount Performance:

In [95]:
df["discount_tier"] = pd.cut(
    df["discount_percentage"],
    bins=[0, 10, 30, 50, 100],
    labels=["0-10%", "10-30%", "30-50%", "50%+"],
    include_lowest=True
)

df["discount_tier"].value_counts()

C:\Users\Sahil Khan\AppData\Local\Temp\ipykernel_22980\3439198061.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



discount_tier
50%+      693
30-50%    426
10-30%    259
0-10%      84
Name: count, dtype: int64

In [96]:
tier_analysis = (
    df.groupby("discount_tier", observed=False)
      .agg(
          Avg_Rating_Count=("rating_count", "mean"),
          Avg_Revenue=("Revenue_Potential", "mean"),
          Avg_Discount=("discount_percentage", "mean"),
          Product_Count=("product_name", "count")
      )
      .reset_index()
)

# Convert to Thousands (K)
tier_analysis["Avg_Revenue"] = (tier_analysis["Avg_Revenue"] / 1000).round(2)

tier_analysis

,discount_tier,Avg_Rating_Count,Avg_Revenue,Avg_Discount,Product_Count
0,0-10%,12086.261905,55580.45,2.809524,84
1,10-30%,21457.169884,132503.10,22.166023,259
2,30-50%,16648.633803,51312.87,41.934272,426
3,50%+,18903.916306,15975.98,66.170274,693


In [97]:
comparison = tier_analysis.melt(
    id_vars="discount_tier",
    value_vars=["Avg_Rating_Count", "Avg_Revenue"],
    var_name="Metric",
    value_name="Value"
)

fig = px.bar(
    comparison,
    x="discount_tier",
    y="Value",
    color="Metric",
    barmode="group",
    title="Discount Tier Performance"
)

fig.update_layout(template="plotly_white")

fig.show()

##### 1.0–10% Discount Tier
###### Very low discounts generate reasonable revenue, but customer engagement remains lower than products receiving moderate discounts.

##### 2.10–30% Discount Tier ⭐ (Best Performing)
###### Products with moderate discounts (10–30%) achieve the best balance between customer demand and revenue generation. This appears to be the most effective pricing strategy.

##### 3.30–50% Discount Tier
###### Increasing discounts beyond 30% does not improve customer engagement and is associated with lower revenue potential.

##### 4.50%+ Discount Tier
###### Deep discounts (50%+) do not generate proportionally higher customer engagement and are associated with the lowest revenue potential, suggesting that excessive discounting may be reducing profitability.



# Why is 10–30% Better than 50%?

## Suppose 
Product A

Original Price = £100

Discount = 20%

Selling Price = £80

Rating Count = 2000

Revenue Potential:
80 × 2000
=160000

Product B

Original Price = £100

Discount = 60%

Selling Price = £40

Rating Count = 2300

Revenue Potential:
40 ×2300
=92000

See Customers Only increase 
2000 -> 23000 
But Price Decreasee 
80 -> 40 (Revenue almost half.) leads to (Over Discounting.)



## Product Distribution

Interesting Things.

Tier Products

1.0–10%	84

2.10–30%	259

3.30–50%	426

4.50%+	693

Almost Half Dataset Covers in 50% Discount Means Company Gives Discount Aggressively But The Product Above 50% Gives Least Reveneue so we need to improve Pricing strategy

# 2. Profitability & Revenue Realization

#### Discount  Leakage

##### Step 1: Total Discount Leakage

In [98]:
df["Total_Discount_Leakage"] = (
    df["absolute_discounted_amount"] * df["rating_count"]
)

C:\Users\Sahil Khan\AppData\Local\Temp\ipykernel_22980\1183686616.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



##### Step 2:Discount Leakage by Categorye

In [99]:
category_discount = (
    df.groupby("category")["Total_Discount_Leakage"]
      .sum()
      .sort_values(ascending=False)
      .reset_index()
)

# Convert to Lakhs
category_discount["Total_Discount_Leakage_Lakh"] = (
    category_discount["Total_Discount_Leakage"] / 100000
).round(2)

category_discount

,category,Total_Discount_Leakage,Total_Discount_Leakage_Lakh
0,Electronics,3.884105e+10,388410.53
1,Computers&Accessories,6.260240e+09,62602.40
2,Home&Kitchen,4.194989e+09,41949.89
3,MusicalInstruments,8.665899e+07,866.59
4,OfficeProducts,1.499209e+07,149.92
5,Health&PersonalCare,3.666663e+06,36.67
6,HomeImprovement,3.576244e+06,35.76
7,Car&Motorbike,1.856998e+06,18.57
8,Toys&Games,0.000000e+00,0.00


In [100]:

fig = px.bar(
    category_discount,
    x="category",
    y="Total_Discount_Leakage",
    title="Total Discount Leakage by Category",
    color="Total_Discount_Leakage",
    text_auto=".2s"
)

fig.update_layout(
    template="plotly_white",
    xaxis_title="Category",
    yaxis_title="Total Discount Amount (£)"
)

fig.show()

##### Step 3: Top 10 Products with Highest Discount Leakage

In [101]:
top_discount = (
    df.sort_values("Total_Discount_Leakage", ascending=False)
      .head(10)
)

top_discount

,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count,Revenue_Potential,absolute_discounted_amount,discount_tier,Total_Discount_Leakage
568,B08VB57558,"Samsung Galaxy S20 FE 5G (Cloud Navy, 8GB RAM,...",Electronics,37990.0,74999.0,49.0,4.2,27790.0,1.055742e+09,37009.0,30-50%,1.028480e+09
935,B08GTYFC37,"SanDisk 1TB Extreme Portable SSD 1050MB/s R, 1...",Computers&Accessories,10389.0,32000.0,68.0,4.4,41398.0,4.300838e+08,21611.0,50%+,8.946522e+08
370,B09GFPVD9Y,"Redmi 9 Activ (Carbon Black, 4GB RAM, 64GB Sto...",Electronics,8499.0,10999.0,23.0,4.1,313836.0,2.667292e+09,2500.0,10-30%,7.845900e+08
72,B09RFC46VP,Redmi 108 cm (43 inches) 4K Ultra HD Android S...,Electronics,26999.0,42999.0,37.0,4.2,45238.0,1.221381e+09,16000.0,30-50%,7.238080e+08
371,B09GFLXVH9,"Redmi 9A Sport (Coral Green, 2GB RAM, 32GB Sto...",Electronics,6499.0,8499.0,24.0,4.1,313836.0,2.039620e+09,2000.0,10-30%,6.276720e+08
566,B09GFPN6TP,"Redmi 9A Sport (Coral Green, 3GB RAM, 32GB Sto...",Electronics,7499.0,9499.0,21.0,4.1,313832.0,2.353426e+09,2000.0,10-30%,6.276640e+08
124,B08Y55LPBF,Redmi 126 cm (50 inches) 4K Ultra HD Android S...,Electronics,32999.0,44999.0,27.0,4.2,45238.0,1.492809e+09,12000.0,10-30%,5.428560e+08
64,B09F9YQQ7B,Redmi 80 cm (32 inches) Android 11 Series HD R...,Electronics,13999.0,24999.0,44.0,4.2,45238.0,6.332868e+08,11000.0,30-50%,4.976180e+08
953,B09F9YQQ7B,Redmi 80 cm (32 inches) Android 11 Series HD R...,Electronics,13999.0,24999.0,44.0,4.2,45237.0,6.332728e+08,11000.0,30-50%,4.976070e+08
332,B09RWQ7YR6,MI 138.8 cm (55 inches) 5X Series 4K Ultra HD ...,Electronics,46999.0,69999.0,33.0,4.3,21252.0,9.988227e+08,23000.0,30-50%,4.887960e+08


##### Step 4: Compare Discount vs Revenue

In [102]:
comparison = top_discount.melt(
    id_vars="product_name",
    value_vars=["Total_Discount_Leakage", "Revenue_Potential"],
    var_name="Metric",
    value_name="Value"
)

fig = px.bar(
    comparison,
    x="product_name",
    y="Value",
    color="Metric",
    barmode="group",
    title="Top 10 Products: Discount Leakage vs Revenue Potential",
    text_auto=".2s"
)

fig.update_layout(
    template="plotly_white",
    xaxis_title="Product",
    yaxis_title="Amount (£)"
)

fig.update_xaxes(tickangle=45)

fig.show()

##### Step 5: Discount Efficiency Ratio 

In [103]:
df["Discount_Efficiency"] = (
    df["Revenue_Potential"] /
    df["Total_Discount_Leakage"]
).round(2)

C:\Users\Sahil Khan\AppData\Local\Temp\ipykernel_22980\924716093.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [106]:
lowest_efficiency = (
    df.sort_values("Discount_Efficiency")
      .loc[:, [
          "product_name",
          "category",
          "absolute_discounted_amount",
          "Revenue_Potential",
          "Discount_Efficiency"
      ]]
      .head(10)
)

lowest_efficiency

,product_name,category,absolute_discounted_amount,Revenue_Potential,Discount_Efficiency
695,"rts [2 Pack] Mini USB C Type C Adapter Plug, T...",Computers&Accessories,4705.0,1301244.0,0.06
372,"Fire-Boltt Ninja Call Pro Plus 1.83"" Smart Wat...",Electronics,18200.0,25072663.0,0.10
364,"Fire-Boltt Ninja Call Pro Plus 1.83"" Smart Wat...",Electronics,18200.0,25072663.0,0.10
368,"Fire-Boltt Ninja Call Pro Plus 1.83"" Smart Wat...",Electronics,18200.0,25072663.0,0.10
380,"Fire-Boltt Ninja Call Pro Plus 1.83"" Smart Wat...",Electronics,18200.0,25072663.0,0.10
334,"Fire-Boltt Ninja Call Pro Plus 1.83"" Smart Wat...",Electronics,18200.0,25072663.0,0.10
644,Sounce Spiral Charger Cable Protector Data Cab...,Computers&Accessories,900.0,138204.0,0.11
407,Sounce Spiral Charger Cable Protector Data Cab...,Computers&Accessories,900.0,138204.0,0.11
920,"Silicone Rubber Earbuds Tips, Eartips, Earpads...",Electronics,900.0,58806.0,0.11
602,LAPSTER Spiral Charger Spiral Charger Cable Pr...,Computers&Accessories,900.0,866349.0,0.11
